File: 백준/Silver/4949. 균형잡힌 세상/균형잡힌 세상.cc
#include<bits/stdc++.h> 
using namespace std;
string s;
int main()
{
	ios_base::sync_with_stdio(false);
	cin.tie(NULL); cout.tie(NULL);
	
	while(true)
	{
		stack<char> stk;
		getline(cin,s);
		if(s==".") break;
		for(int i=0;i<s.size();i++)
		{
			if(s[i]=='(' || s[i]=='[')
			{
				stk.push(s[i]);	
			}
			else if(s[i]==')')
			{
				if(!stk.empty() && stk.top()=='(')
				{
					stk.pop();
				}
				else
				{
					stk.push(s[i]);
				}
			}
			else if(s[i]==']')
			{
				if(!stk.empty() && stk.top()=='[')
				{
					stk.pop();
				}
				else
				{
					stk.push(s[i]);
				}
			}
		}
		if(stk.size()==0) cout<<"yes"<<"\n";
		else cout<<"no"<<"\n";
	} 
	
	return 0;
}

File: 백준/Silver/4949. 균형잡힌 세상/균형잡힌 세상.py
while True:
    line = input()
    if line == '.':
        break
    stack=[]
    for ch in line:
        if ch in ['(', '[']:
            stack.append(ch)
        elif ch==')':
            if stack and stack[-1]=='(':
                stack.pop()
            else:
                stack.append(ch)
        elif ch==']':
            if stack and stack[-1]=='[':
                stack.pop()
            else:
                stack.append(ch)
            
    if not stack:
        print('yes')
    else:
        print('no')